### PR TITLE
Add experimental multimap insert and contains

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds a new `experimental::static_multimap` using the new OA design. It adds `insert` and `contains` operations for the new multimap.

### CODE REVIEW

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 1]
```diff
@@ -25,6 +25,280 @@
...
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::clear(
+  cuda::stream_ref stream)
+{
+  impl_->clear(stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl:128]:
```suggestion
  this->impl_->clear(stream);
```

Not a hard requirement but I think the RAPIDS best practice guide mentions using `this->` when applicable

File: include/cuco/static_multimap.cuh [hunk 2]
```diff
@@ -39,6 +42,364 @@
 #include <utility>
 
 namespace cuco {
+namespace experimental {
+/**
+ * @brief A GPU-accelerated, unordered, associative container of key-value pairs that supports
+ * equivalent keys.
+ *
+ * The `static_map` supports two types of operations:
```
Comment (sleeepyjack) [inline include/cuco/static_multimap.cuh]:
```suggestion
 * The `static_multimap` supports two types of operations:
```

**Review by sleeepyjack — COMMENTED** [review]
First round.

Looks great!

It's very pleasant to see that with all of our previous refactoring work, implementing the new `static_multimap` boils down simply dispatching from the frontend API to the OA backend API ❤️

File: include/cuco/detail/static_multimap/static_multimap.inl (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/static_multimap/static_multimap.inl:128]: Ah, we do use `this->` for member functions. For data members with `_` suffix, the intention seems to be clear enough.

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 3]
```diff
@@ -25,6 +25,280 @@
...
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+constexpr static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::
+  mapped_type
+  static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::
+    empty_value_sentinel() const noexcept
+{
+  return this->empty_value_sentinel_;
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
  return empty_value_sentinel_;
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+                                                  T,
+                                                  Scope,
+                                                  KeyEqual,
+                                                  ProbingScheme,
+                                                  StorageRef,
+                                                  Operators...>::key_equal
+static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::key_eq()
+  const noexcept
+{
+  return this->impl_.key_eq();
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
  return impl_.key_eq();
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+                                                  T,
+                                                  Scope,
+                                                  KeyEqual,
+                                                  ProbingScheme,
+                                                  StorageRef,
+                                                  Operators...>::const_iterator
+static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::end()
+  const noexcept
+{
+  return this->impl_.end();
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
  return impl_.end();
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+  KeyEqual,
+  ProbingScheme,
+  StorageRef,
+  Operators...>::with_operators(NewOperators...) const noexcept
+{
+  return static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, NewOperators...>{
+    cuco::empty_key<Key>{this->empty_key_sentinel()},
+    cuco::empty_value<T>{this->empty_value_sentinel()},
+    this->key_eq(),
+    this->impl_.probing_scheme(),
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
    impl_.probing_scheme(),
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+    {},
+    this->impl_.storage_ref()};
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
    impl_.storage_ref()};
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+          typename StorageRef,
+          typename... Operators>
+template <typename CG, cuda::thread_scope NewScope>
+__device__ constexpr auto
+static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::make_copy(
+  CG const& tile,
+  window_type* const memory_to_use,
+  cuda_thread_scope<NewScope> scope) const noexcept
+{
+  this->impl_.make_copy(tile, memory_to_use);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
  impl_.make_copy(tile, memory_to_use);
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+  return static_multimap_ref<Key, T, NewScope, KeyEqual, ProbingScheme, StorageRef, Operators...>{
+    cuco::empty_key<Key>{this->empty_key_sentinel()},
+    cuco::empty_value<T>{this->empty_value_sentinel()},
+    cuco::erased_key<Key>{this->erased_key_sentinel()},
+    this->key_eq(),
+    this->impl_.probing_scheme(),
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
    impl_.probing_scheme(),
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+          typename KeyEqual,
+          typename ProbingScheme,
+          typename StorageRef,
+          typename... Operators>
+template <typename CG>
+__device__ constexpr void
+static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::initialize(
+  CG const& tile) noexcept
+{
+  this->impl_.initialize(tile);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]:
```suggestion
  impl_.initialize(tile);
```

File: include/cuco/detail/static_multimap/static_multimap_ref.inl [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+ */
+
+#pragma once
+
+#include <cuco/detail/equal_wrapper.cuh>
+#include <cuco/operator.hpp>
+
+#include <cuda/atomic>
+#include <cuda/std/functional>
+#include <thrust/tuple.h>
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap_ref.inl]: where do we use this header?

File: include/cuco/static_multimap_ref.cuh [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+      static_multimap_ref<Key, T, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>>... {
+  /// Flag indicating whether duplicate keys are allowed or not
+  static constexpr auto allows_duplicates = true;
+
+  /// Implementation type
+  using impl_type = detail::
+    open_addressing_ref_impl<Key, Scope, KeyEqual, ProbingScheme, StorageRef, allows_duplicates>;
+
+  static_assert(sizeof(T) == 4 or sizeof(T) == 8,
+                "sizeof(mapped_type) must be either 4 bytes or 8 bytes.");
```
Comment (sleeepyjack) [inline include/cuco/static_multimap_ref.cuh:78]: I thought `cuda::atomic` would support 2 byte types by now

**Review by sleeepyjack — APPROVED** [review]
Apart from a few minor thinsg this PR is good to go. I'm gonna approve it now so we can merge it ASAP.

File: include/cuco/static_multimap_ref.cuh (hunk 12 above)
Comment (PointKernel) [reply include/cuco/static_multimap_ref.cuh:78]: https://godbolt.org/z/a78x4qrha wow, good to know. This unblocks many libcudf cleanups as well.

File: include/cuco/detail/static_multimap/static_multimap_ref.inl (hunk 11 above)
Comment (PointKernel) [reply include/cuco/detail/static_multimap/static_multimap_ref.inl]: Good catch!

File: include/cuco/static_multimap_ref.cuh (hunk 12 above)
Comment (PointKernel) [reply include/cuco/static_multimap_ref.cuh:78]: I will create a separate PR to address this.

Comment (PointKernel) [reply include/cuco/static_multimap_ref.cuh:78]: https://github.com/NVIDIA/cuCollections/issues/554

**Pull request merged**